tests/testthat/test-paste_linter.R

```diff
@@ -113,3 +113,95 @@ test_that("paste_linter blocks simple disallowed usages", {
   expect_lint("paste0(rep('*', 20L), collapse='')", lint_msg, linter)
   expect_lint("paste(rep('#', width), collapse='')", lint_msg, linter)
 })
+
+test_that("paste_linter skips allowed usages for file paths", {
+  linter <- paste_linter()
+
+  expect_lint("paste('a', 'b', 'c')", NULL, linter)
+  expect_lint("paste('a', 'b', 'c', sep = ',')", NULL, linter)
+  expect_lint("paste('a', 'b', collapse = '/')", NULL, linter)
+  expect_lint("cat(paste('a', 'b'), sep = '/')", NULL, linter)
+  expect_lint("sep <- '/'; paste('a', sep)", NULL, linter)
+  expect_lint("paste(sep = ',', '/', 'a')", NULL, linter)
+
+  # paste(..., sep='/', collapse=collapse) is not a trivial swap to file.path
+  expect_lint("paste(x, y, sep = '/', collapse = ':')", NULL, linter)
+
+  expect_lint("file.path('a', 'b', 'c')", NULL, linter)
+
+  # testing the sep starts with / is not enough
+  expect_lint("paste('a', 'b', sep = '//')", NULL, linter)
+})
+
+test_that("paste_linter blocks simple disallowed usages for file paths", {
+  linter <- paste_linter()
+  lint_msg <- rex::rex("Construct file paths with file.path(...) instead of")
+
+  expect_lint("paste(sep = '/', 'a', 'b')", lint_msg, linter)
+  expect_lint("paste('a', 'b', sep = '/')", lint_msg, linter)
+})
+
+test_that("paste_linter ignores non-path cases with paste0", {
+  linter <- paste_linter()
+
+  expect_lint("paste0(x, y)", NULL, linter)
+  expect_lint("paste0('abc', 'def')", NULL, linter)
+  expect_lint("paste0('/abc', 'def/')", NULL, linter)
+  expect_lint("paste0(x, 'def/')", NULL, linter)
+  expect_lint("paste0('/abc', y)", NULL, linter)
+  expect_lint("paste0(foo(x), y)", NULL, linter)
+  expect_lint("paste0(foo(x), 'def')", NULL, linter)
+
+  # these might be a different lint (as.character instead, e.g.) but not here
+  expect_lint("paste0(x)", NULL, linter)
+  expect_lint("paste0('a')", NULL, linter)
+  expect_lint("paste0('a', 1)", NULL, linter)
+
+  # paste0(..., collapse=collapse) not directly mapped to file.path
+  expect_lint("paste0(x, collapse = '/')", NULL, linter)
+})
+
+test_that("paste_linter detects paths built with '/' and paste0", {
+  linter <- paste_linter()
+  lint_msg <- rex::rex("Construct file paths with file.path(...) instead of")
+
+  expect_lint("paste0(x, '/', y)", lint_msg, linter)
+  expect_lint("paste0(x, '/', y, '/', z)", lint_msg, linter)
+  expect_lint("paste0(x, '/abc/', 'def/', y)", lint_msg, linter)
+  expect_lint("paste0(foo(x), '/abc/', 'def/', bar(y))", lint_msg, linter)
+})
+
+test_that("paste_linter skips initial/terminal '/' and repeated '/' for paths", {
+  linter <- paste_linter()
+
+  expect_lint("paste0('/', x)", NULL, linter)
+  expect_lint("paste0(x, '/')", NULL, linter)
+  expect_lint("paste0(x, '//hey/', y)", NULL, linter)
+  expect_lint("paste0(x, '/hey//', y)", NULL, linter)
```

Comment:
Seeing this test I wonder if we can force a false positive with `paste0(a, "/not//a/path/", b)`?

Reply:
whether that's a path is a bit ambiguous (it's a valid path on many OS), and would also disallow it being used to construct URLs... maybe that's another option to add? Since I know some (maybe many) authors don't like building URLs with `file.path()`, we could disable the linter if `//` is present? Or should we just do that by default

Reply:
A default-TRUE argument to ignore URLs sounds good.

Reply:
In the spirit of importing something closer to the original linter we used internally first, marked this piece for follow-up: #2082. It's part of the release milestone to ensure this is the default behavior before the next CRAN version.